### PR TITLE
Fix deferring parameters for class::localparam access (#7746)

### LINKED ISSUE #7746: CFG::param usage in module parameters during elaboration
@em2machine Same issue as #7609 still occurs with module instantiations (or interfaces):

The following fails
```systemverilog
module Submodule #(parameter int width) (); endmodule

module test ();
    virtual class C #(parameter int a);
        localparam int b = a;
    endclass

    typedef C#(8) CFGInst;

    localparam int b = CFGInst::b;

    Submodule #(b) intf();
endmodule
```

with the error
```
%Error-UNSUPPORTED: test.sv:10:33: dotted expressions in parameters
                                 : ... note: In instance 'test'
                                 : ... Suggest use a typedef
   10 |     localparam int b = CFGInst::b;
      |                                 ^
                    ... For error description see https://verilator.org/warn/UNSUPPORTED?v=5.049
%Error: Internal Error: test.sv:10:33: ../V3Width.cpp:8733: Node has no type
                                     : ... note: In instance 'test'
   10 |     localparam int b = CFGInst::b;
      |                                 ^
                        ... This fatal error may be caused by the earlier error(s); resolve those first.
```

Note here that directly using the parameter works fine:
```systemverilog
module Submodule #(parameter int width) (); endmodule

module test ();
    virtual class C #(parameter int a);
        localparam int b = a;
    endclass

    typedef C#(8) CFGInst;

    Submodule #(CFGInst::b) intf();
endmodule
```

Looks like the same class of issue where we need to defer usages.

### PR DESCRIPTION
**Pull request by UnsignedByte**
Fixes #7746 

Tests were mostly done with claude to get a set of mvps up, more as a scaffold for what we might want to land. The set of tests here should cover most of the remaining edge cases

Fixes:
* Module instantiations: `localparam p = C::b; Submodule #(.p(p)) ();`
* Multilevel class chains (where param C uses class B which uses class A in its definition)
* `Class::member.field` accesses
* Dependent typedefs `typedef logic [C#(val)::b-1:0] type_t`

### CODE REVIEW

**Comment by UnsignedByte** [comment]
@wsnyder This works for my uses but isn't particularly clean. This is a structural limitation of parameterization, what we should really be doing here is building a proper topological sort of modules/typedefs/params/classes and then walking the list in the right order and deparameterizing as we go up. I don't think I have the time or will to do a full rewrite though.

If you have any other ideas please let me know

File: src/V3Param.cpp [hunk 1]
```diff
@@ -2178,6 +2232,57 @@ class ParamProcessor final {
...
+                if (VN_IS(dotp->lhsp(), ClassOrPackageRef)) dotps.push_back(dotp);
+            });
+            p->foreach([&](const AstRefDType* refp) {
+                AstTypedef* const tdefp = refp->typedefp();
+                if (tdefp && reachedTypedefs.insert(tdefp).second) {
+                    tdefps.push_back(tdefp);
+                    if (tdefp->subDTypep()) worklist.push_back(tdefp->subDTypep());
+                }
+            });
+            p->foreach([&](const AstVarRef* refp) {
```
Comment (wsnyder) [inline src/V3Param.cpp]: This is recursing the same structure 3 times which is expensive, should be just once.

But is this process duplicating logic that is already implemented elsewhere?

**Comment by wsnyder** [comment]
@em2machine can you please review thanks

**Comment by em2machine** [comment]
Sorry, will review.  

@UnsignedByte Yes, agreed.  I started this in a sandbox (building a complete dependency graph (V3DepGraph) with an OOO graph 'execution' model for robustness and future proofing, etc), but ran out of cycles.  

I may come back to it, but not in the near term (sort of insanely busy right now).  Alsoit would essentially replace the current V3Param, V3LinkDot, and part of V3Width and not sure if @wsnyder would want that big of a structural change.

**Comment by em2machine** [comment]
A couple of items.

1. Unless I am mistaken it looks like the struct-member select in visit(AstVar) duplicates much/most of V3Width::memberSelStruct.  The comment also says "Match V3Width::memberSelStruct".  Seems like this is an opportunity for drift. Can we create a shared helper with V3Width (or defer the member-select to V3Width)?
2. Are there performance issues with the deferred-dot resolution running unconditionally at every generate/case node?  The `if (dotps.empty() && tdefps.empty()) return;`  early-out is after the full foreach, so every generate for, if, case in a design pays a subtree walk during V3Param even when zero class params are involved.  It seems like in the extreme case this goes super-linear.

Thanks.

**Comment by UnsignedByte** [comment]
Hey, sorry for the late response, I was on vacation. 
1. I thought about this but I don't think I know enough about V3Width to confidently make a change here. The V3Param instance does some other stuff like resolving struct accesses as well, it doesn't seem to map 1:1 to `memberSelStruct`. Feel free to take over this branch and make any changes though that you deem fit.
2. This should be no different performance wise as the previous change already does the foreach loop. I don't see an easy way to detect whether a node depends on class params/etc without doing this, unless we want to break this out into its own pass somehow.

Once again feel free to override/make any changes you guys want though!

**Comment by em2machine** [comment]
Ok, I may take a look here in a couple of days.  My plan is to start looking at the larger structural (architectural) change again in the middle of August.  I would build it alongside the existing infrastructure so I can run both in parallel looking / understanding divergence, etc.  But I won't be able to start that / pick up my previous work until that timeframe.

File: src/V3Param.cpp [hunk 2]
```diff
@@ -1288,10 +1288,59 @@ class ParamProcessor final {
             // Param/lparam member: substitute its constant value so the caller's constify can
             // succeed.
             if (varp->isParam() && varp->valuep()) {
-                if (!VN_IS(varp->valuep(), Const)) V3Const::constifyParamsEdit(varp);
+                if (!VN_IS(varp->valuep(), Const)) {
+                    // If the value contains nested class::member Dots (e.g. this
+                    // member's expression references another paramed class), resolve
+                    // those first so constify can fold the whole chain.
+                    std::vector<AstDot*> nestedDots;
+                    varp->valuep()->foreach([&](AstDot* dp) {
```
Comment (RRozak) [inline src/V3Param.cpp]:
```suggestion
                    varp->valuep()->foreach([&](AstDot* const dp) {
```

File: src/V3Param.cpp [hunk 3]
```diff
@@ -1288,10 +1288,59 @@ class ParamProcessor final {
...
-                    VL_DO_DANGLING(dotp->deleteTree(), dotp);
+                    // Walk up any outer `.fieldN[.fieldM...]` chain (struct lparam
+                    // member access like `CFG::cfg.jt.cam_type`) and accumulate the
+                    // packed-struct bit offset.  Replace the topmost Dot with a
+                    // Sel(constp, lsb, width) matching V3Width::memberSelStruct.
+                    int totalLsb = 0;
+                    int sliceWidth = varp->width();
+                    AstNodeDType* curDTypep = varp->dtypep();
+                    AstNode* topp = dotp;
+                    AstDot* outerDot = VN_CAST(dotp->backp(), Dot);
```
Comment (RRozak) [inline src/V3Param.cpp]:
```suggestion
                    AstDot* outerDotp = VN_CAST(dotp->backp(), Dot);
```

File: src/V3Param.cpp [hunk 4]
```diff
@@ -1288,10 +1288,59 @@ class ParamProcessor final {
...
+                    while (outerDot && outerDot->lhsp() == topp) {
+                        const AstParseRef* const fieldRef = VN_CAST(outerDot->rhsp(), ParseRef);
```
Comment (RRozak) [inline src/V3Param.cpp]:
```suggestion
                        const AstParseRef* const fieldRefp = VN_CAST(outerDot->rhsp(), ParseRef);
```

File: src/V3Param.cpp [hunk 5]
```diff
@@ -1288,10 +1288,59 @@ class ParamProcessor final {
...
+                        if (!fieldRef) break;
+                        AstNodeDType* const skippedp = curDTypep ? curDTypep->skipRefp() : nullptr;
+                        AstNodeUOrStructDType* const structp
+                            = VN_CAST(skippedp, NodeUOrStructDType);
+                        if (!structp) break;
+                        AstMemberDType* const foundMemp = VN_CAST(
+                            m_memberMap.findMember(structp, fieldRef->name()), MemberDType);
+                        if (!foundMemp) break;
```
Comment (RRozak) [inline src/V3Param.cpp]: Is this case possible? If not, please use `UASSERT_OBJ` instead.

File: src/V3Param.cpp [hunk 6]
```diff
@@ -1288,10 +1288,59 @@ class ParamProcessor final {
...
+                        dotp->replaceWith(constp->cloneTree(false));
+                        VL_DO_DANGLING(dotp->deleteTree(), dotp);
+                    } else {
+                        AstConst* const clonep = static_cast<AstConst*>(constp->cloneTree(false));
+                        AstSel* const selp
+                            = new AstSel{topp->fileline(), clonep, totalLsb, sliceWidth};
+                        // Match V3Width::memberSelStruct: skip RefDTypes to surface
+                        // enum dtype, and mark didWidth so V3Width doesn't reflatten.
+                        if (curDTypep) selp->dtypep(curDTypep->skipRefToEnump());
+                        selp->didWidth(true);
```
Comment (RRozak) [inline src/V3Param.cpp]: Can `curDTypep` be nullptr? If so, we shouldn't set `didWidth` to true if it's dtypep is unknown.

File: src/V3Param.cpp [hunk 7]
```diff
@@ -2178,6 +2227,57 @@ class ParamProcessor final {
...
+    // are also kept in lists for the resolution passes at the end.
+    void resolveDeferredDotsReachableFrom(AstNode* rootp, AstNodeModule* modp) {
+        std::vector<AstDot*> dotps;
+        std::vector<AstTypedef*> tdefps;
+        const auto& deferredVarps = v3Global.rootp()->deferredParamVarps();
+        std::set<AstVar*> reachedDeferred;
+        std::set<const AstTypedef*> reachedTypedefs;
+        std::vector<AstNode*> worklist{rootp};
+        while (!worklist.empty()) {
+            AstNode* const p = worklist.back();
```
Comment (RRozak) [inline src/V3Param.cpp]: Please use a longer, self-descriptive name.

File: src/V3Param.cpp [hunk 8]
```diff
@@ -2178,6 +2227,57 @@ class ParamProcessor final {
...
+            worklist.pop_back();
+            p->foreach([&](AstNode* np) {
```
Comment (RRozak) [inline src/V3Param.cpp]:
```suggestion
            p->foreach([&](AstNode* const np) {
```

File: src/V3Param.cpp [hunk 9]
```diff
@@ -2178,6 +2227,57 @@ class ParamProcessor final {
...
+                if (AstDot* const dotp = VN_CAST(np, Dot)) {
+                    if (VN_IS(dotp->lhsp(), ClassOrPackageRef)) dotps.push_back(dotp);
+                } else if (const AstRefDType* const refp = VN_CAST(np, RefDType)) {
+                    AstTypedef* const tdefp = refp->typedefp();
+                    if (tdefp && reachedTypedefs.insert(tdefp).second) {
+                        tdefps.push_back(tdefp);
+                        if (tdefp->subDTypep()) worklist.push_back(tdefp->subDTypep());
```
Comment (RRozak) [inline src/V3Param.cpp]: Typedefs should always have non-null `subDTypep()`: https://github.com/verilator/verilator/blob/506f703da5ec43ae1709d90b952f50fd36bfcd5e/src/V3AstNodeOther.h#L2018-L2020

File: src/V3Param.cpp [hunk 10]
```diff
@@ -2178,6 +2227,57 @@ class ParamProcessor final {
...
+                    }
+                } else if (const AstVarRef* const refp = VN_CAST(np, VarRef)) {
+                    AstVar* const varp = refp->varp();
+                    if (varp && varp->varType() == VVarType::LPARAM && deferredVarps.count(varp)
```
Comment (RRozak) [inline src/V3Param.cpp]: VarRefs always have non-null `varp()`
```suggestion
                    if (varp->varType() == VVarType::LPARAM && deferredVarps.count(varp)
```

File: src/V3Param.cpp [hunk 11]
```diff
@@ -2178,6 +2227,57 @@ class ParamProcessor final {
...
+                        && reachedDeferred.insert(varp).second && varp->valuep()) {
```
Comment (RRozak) [inline src/V3Param.cpp]: I think that parameters should always have non-null `valuep()`. Please replace this condition with an assert.

**Review by RRozak — CHANGES_REQUESTED** [review]

File: src/V3Param.cpp (hunk 11 above)
Comment (wsnyder) [reply src/V3Param.cpp]: I don't see the assert in the updates.

Comment (wsnyder) [reply src/V3Param.cpp]: (Should be a `UASSERT_OBJ(varp->valuep(), varp, ...`)

**Comment by RRozak** [comment]
> what we should really be doing here is building a proper topological sort of modules/typedefs/params/classes and then walking the list in the right order and deparameterizing as we go up. I don't think I have the time or will to do a full rewrite though.

We talked about it in https://github.com/verilator/verilator/issues/4391. Building a graph and sort nodes topologically is in my opinion the best approach.

File: src/V3Param.cpp (hunk 5 above)
Comment (RRozak) [reply src/V3Param.cpp]: You haven't added an assert. Is it possible then? Do we have a test covering it?

File: src/V3Param.cpp [hunk 12]
```diff
@@ -1288,10 +1288,65 @@ class ParamProcessor final {
...
+                    AstDot* outerDotp = VN_CAST(dotp->backp(), Dot);
+                    // On any mismatch (non-ParseRef rhs, non-packed-struct dtype, or
+                    // unknown field name) leave the outer Dots intact so V3Width can
+                    // emit the proper user-facing diagnostic.
+                    while (outerDotp && outerDotp->lhsp() == topp) {
+                        const AstParseRef* const fieldRefp = VN_CAST(outerDotp->rhsp(), ParseRef);
+                        if (!fieldRefp) break;
+                        AstNodeDType* const skippedp = curDTypep ? curDTypep->skipRefp() : nullptr;
+                        AstNodeUOrStructDType* const structp
+                            = VN_CAST(skippedp, NodeUOrStructDType);
```
Comment (RRozak) [inline src/V3Param.cpp]: What if a structure is unpacked? A few lines below you are referring to `width` and `lsb` of its members and these values seem to make sense only in the case of packed types.

File: src/V3Param.cpp [hunk 13]
```diff
@@ -1288,10 +1288,65 @@ class ParamProcessor final {
...
+                        dotp->replaceWith(constp->cloneTree(false));
+                        VL_DO_DANGLING(dotp->deleteTree(), dotp);
+                    } else {
+                        AstConst* const clonep = static_cast<AstConst*>(constp->cloneTree(false));
+                        AstSel* const selp
+                            = new AstSel{topp->fileline(), clonep, totalLsb, sliceWidth};
+                        // Match V3Width::memberSelStruct: skip RefDTypes to surface
+                        // enum dtype, and mark didWidth so V3Width doesn't reflatten.
+                        // Only mark didWidth when we have a resolved dtype to attach.
+                        if (curDTypep) {
```
Comment (RRozak) [inline src/V3Param.cpp]: Do we have any test covering it? I think that if `curDTypep`, which is a type of last member, is nullptr, then `totalLsb` and `sliceWidth` aren't correct too, so the `AstSel` node we created is probably incorrect.

**Comment by wsnyder** [comment]
@RRozak thanks for reviewing, can you make another pass, when satisfied and coverage is 100% feel free to merge.

File: src/V3Param.cpp (hunk 10 above)
Comment (RRozak) [reply src/V3Param.cpp]: Unresolving, because you haven't applied this suggestion.

It's great that you are working on the coverage. For now only 3 branches are missing and this is one of them. Generally, it is uncoverable, because varrefs always have varp() set. Please remove it, because it doesn't make sense to check it.

**Review by RRozak — CHANGES_REQUESTED** [review]

**Comment by UnsignedByte** [comment]
Could I get a review on this (@RRozak @wsnyder)

File: src/V3Param.cpp [hunk 14]
```diff
@@ -1250,10 +1250,84 @@ class ParamProcessor final {
...
+                    int totalLsb = 0;
+                    int sliceWidth = varp->width();
+                    AstNodeDType* curDTypep = varp->dtypep();
+                    AstNode* topp = dotp;
+                    AstDot* outerDotp = VN_CAST(dotp->backp(), Dot);
+                    const auto errorRecover = [&](AstDot* const badp, const string& msg) {
+                        badp->v3error(msg);
+                        badp->replaceWith(new AstConst{badp->fileline(), AstConst::Signed32{}, 0});
+                        VL_DO_DANGLING(badp->deleteTree(), badp);
+                    };
```
Comment (wsnyder) [inline src/V3Param.cpp]: Please put this code explicitly instead of using a helper as we have special code that looks for v3error.

File: src/V3Param.cpp [hunk 15]
```diff
@@ -1250,10 +1250,84 @@ class ParamProcessor final {
             // Param/lparam member: substitute its constant value so the caller's constify can
             // succeed.
             if (varp->isParam() && varp->valuep()) {
-                if (!VN_IS(varp->valuep(), Const)) V3Const::constifyParamsEdit(varp);
+                if (!VN_IS(varp->valuep(), Const)) {
+                    // If the value contains nested class::member Dots (e.g. this
+                    // member's expression references another paramed class), resolve
```
Comment (wsnyder) [inline src/V3Param.cpp]: This whole function is huge, let's not make it worse, put the below into a separate function if you can.  And/or can make a pre-PR just to refactor it without other changes.

**Review by wsnyder — CHANGES_REQUESTED** [review]

**Comment by sifferman** [comment]
Thanks for your work on this, guys! I just tested on this PR's branch (Verilator 5.051 devel rev v5.050-167-g86175ed5e), and I believe I found a bug.

This patch still rejects a class-scoped typedef used as a `localparam type` when that type is passed to a child module's `type` parameter. This is valid IEEE 1800-2023 SystemVerilog and is accepted by other simulators and sv2v.

## Test case

```systemverilog
class types_c #(parameter int W = 8);
    typedef logic [W-1:0] t;
endclass

module child #(parameter type T = logic) (input T a_i);
endmodule

module top #(
    parameter  int  W = 8,
    localparam type t = types_c#(W)::t
) ();
    child #(.T(t)) u (.a_i('0)); // error is here
endmodule
```

```
verilator --lint-only class-type-param.sv --top top
```

**Expected:** lints cleanly.

**Actual:**

```
%Error-UNSUPPORTED: class-type-param.sv:10:38: dotted expressions in parameters
                                             : ... note: In instance 'top'
                                             : ... Suggest use a typedef
   10 |     localparam type t = types_c#(W)::t
      |                                      ^
                    ... For error description see https://verilator.org/warn/UNSUPPORTED?v=5.051
%Error: class-type-param.sv:10:38: Expected data type, not a DOT
                                 : ... note: In instance 'top'
   10 |     localparam type t = types_c#(W)::t
      |                                      ^
        ... See the manual at https://verilator.org/verilator_doc.html?v=5.051 for more assistance.
%Error: Internal Error: class-type-param.sv:10:21: ../V3Width.cpp:9695: iterateEditMoveDTypep didn't get width resolution of VOIDDTYPE
                                                 : ... note: In instance 'top'
   10 |     localparam type t = types_c#(W)::t
      |                     ^
                        ... This fatal error may be caused by the earlier error(s); resolve those first.
```

The class-scoped `localparam type` on its own lints fine; the failure only occurs when that type is passed to a child module's `type` parameter (remove the `child` instantiation and it passes).

**Comment by wsnyder** [comment]
@sifferman please confirm this is working for you thanks

**Comment by sifferman** [comment]
Yes! That test cases now passes. However, I found another failing test case:
```systemverilog
class types_c #(
    parameter int W = 8
);
  typedef logic [W-1:0] t;
  typedef logic signed [W-1:0] type1_t;
  typedef type1_t [W-1:0] type2_t;
endclass

module child #(
    parameter type T = logic
) (
    input T a_i
);
endmodule

module t;
  localparam int W = 8;

  // Class-scoped type via a localparam type, passed to a child type parameter.
  localparam type t_param = types_c#(W)::t;
  child #(.T(t_param)) u (.a_i('0));

  // Class-scoped type via a module-body typedef alias of a nested typedef.
  typedef types_c#(W)::type2_t t_alias;
  t_alias v;

  initial begin
    v = '0;
    assert ($bits(t_param) == 8);
    assert ($bits(t_alias) == 64);
    $write("*-* All Finished *-*\n");
    $finish;
  end
endmodule
```

Claude generated a patch for it here if relevant: https://github.com/sifferman/verilator/commit/7758ebb1522fbe47a5f2710792f3c8aa68cb1cec

Thanks!

**Comment by sifferman** [comment]
My mistake, I didn't run the full tests. Claude says this should fix the failing t_typedef_circ_bad test:

```diff
diff --git a/src/V3Param.cpp b/src/V3Param.cpp
index 8fcfaec24..fdb0978f7 100644
--- a/src/V3Param.cpp
+++ b/src/V3Param.cpp
@@ -3024,7 +3024,9 @@ class ParamVisitor final : public VNVisitor {
         // A module-body `typedef C#(Cfg)::t alias` the deferred pin/param walk
         // never reached would survive to V3Width unlinked. If its alias chain is
         // broken, run the same deferred resolution here.
-        if (nodep->typedefp() && !nodep->skipRefOrNullp()) {
+        // Circularity is checked first: skipRefOrNullp() reports a recursive
+        // chain itself, which would pre-empt the circularity diagnostic below.
+        if (nodep->typedefp() && !isCircularType(nodep) && !nodep->skipRefOrNullp()) {
             m_processor.resolveDeferredDotsReachableFrom(nodep, m_modp);
         }
         if (isCircularType(nodep)) {
```

**Comment by sifferman** [comment]
Thanks! This now passes all of my internal testbenches

File: src/V3LinkDot.cpp [hunk 16]
```diff
@@ -6089,8 +6139,8 @@ class LinkDotResolveVisitor final : public VNVisitor {
                         << cpackagerefp->warnMore() + "... Suggest '.' instead of '::'");
                 }
             } else {
-                cpackagep->v3warn(E_UNSUPPORTED,
-                                  "Unsupported: Multiple '::' package/class reference");
+                UASSERT_OBJ(scopeDotHandled, cpackagep,
+                            "Unexpected package/class scope operand");
```
Comment (wsnyder) [inline src/V3LinkDot.cpp]: Claude feedback (maybe wrong)

 This replaces a graceful `v3warn(E_UNSUPPORTED, "Unsupported: Multiple
 '::' package/class reference")` with `UASSERT_OBJ(scopeDotHandled, ...)`. `reduceScopeDot()` can return `false` without reporting any error when a *middle* segment of the chain resolves successfully but its `classOrPackageSkipp()` returns null (e.g. a typedef alias of an as-yet-unspecialized parameterized class used as a middle scope, not the last one -- see `t_class_scope_pkg_alias.v`'s comment for why `classOrPackageSkipp()` walks to null in that situation). That looks like it reaches this assert on legal SV, not just an internal invariant violation. Can you confirm whether a 3+-level chain with a deferred/aliased middle segment is actually unreachable here, or should this stay a `v3warn(E_UNSUPPORTED, ...)`?

File: src/V3LinkDot.cpp [hunk 17]
```diff
@@ -5981,6 +5985,44 @@ class LinkDotResolveVisitor final : public VNVisitor {
...
+        for (size_t i = 0; i < refps.size(); ++i) {
+            AstClassOrPackageRef* const refp = refps[i];
+            if (!refp->classOrPackageSkipp() && !refp->classOrPackageNodep()
+                && !m_statep->resolveClassOrPackage(scopeSymp, refp, i == 0, false,
+                                                    "class/package reference")) {
+                return true;  // Error already reported
+            }
+            if (i + 1 == refps.size()) break;
+            AstNodeModule* const modp = refp->classOrPackageSkipp();
+            if (!modp) return false;
```
Comment (wsnyder) [inline src/V3LinkDot.cpp]: Claude feedback (maybe wrong)

* The `if (!modp) return false;` early return (for a non-final segment) reports no diagnostic before returning `false`. If the caller's assertion above is meant to hold, this path needs to either succeed, or itself report a graceful error before returning `false` -- right now it does neither, silently deferring the "was this reachable" question to the caller's assert.

**Review by wsnyder — CHANGES_REQUESTED** [review]
